### PR TITLE
harden AtLeastOnceDeliveryCrashSpec

### PR DESCRIPTION
**Pull request by octonato**
References #31206

### CODE REVIEW

File: akka-persistence/src/test/scala/akka/persistence/AtLeastOnceDeliveryCrashSpec.scala [hunk 1]
```diff
@@ -72,7 +77,9 @@ class AtLeastOnceDeliveryCrashSpec
         system.actorOf(Props(new StoppingStrategySupervisor(testProbe.ref)), "supervisor")
       val superVisor = createCrashActorUnderSupervisor()
       superVisor ! CrashingActor.Message
-      testProbe.expectMsgType[CrashingActor.SendingMessage]
+      val msg = testProbe.expectMsgType[CrashingActor.SendingMessage]
+      // simulates confirmation from probe
+      superVisor ! CrashingActor.Confirmed(msg.deliveryId)
```
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-persistence/src/test/scala/akka/persistence/AtLeastOnceDeliveryCrashSpec.scala:82]: A bit hard to follow test but a quick look at https://github.com/akka/akka/issues/16104 it covers that a message in flight (unconfirmed I guess) isn't replayed on recovery (or even stop). I think it _shouldn't_ confirm or that bugfix is no longer covered.

[the remark this answers is not in the document]
Comment (octonato) [reply akka-persistence/src/test/scala/akka/persistence/AtLeastOnceDeliveryCrashSpec.scala:82]: I don't know if I follow your comment, but what I understand from this test is the following: 

1. we send two message: `CrashingActor.Message` and `CrashingActor.CrashMessage`
2. they won't crash the actor despite the name of the second message
3.`CrashingActor.Message` triggers a `deliver` call
4. actor is explicitly stopped
5. on recover, first message `CrashingActor.Message` is replayed and re-delivered because never confirmed
6. `CrashingActor.CrashMessage` is then replayed causing the actor to fail

I agree that my 'fix' is not helping here. As I understand it now, the test is bogus. 

Is my understand correct that what we should do instead is the following: 

1. send `CrashingActor.Message` and trigger a `deliver` (unconfirmed)
2. send `CrashingActor.CrashMessage` and crash the actor
3. restart the actor, replay `CrashingActor.Message` and confirm that the `deliver` is not retriggered

Is that how you understand it?

[the remark this answers is not in the document]
Comment (octonato) [reply akka-persistence/src/test/scala/akka/persistence/AtLeastOnceDeliveryCrashSpec.scala:82]: Oh, now I see something else

The re-delivery should not happen during recovery, only after full recovery. Since we crash the actor during recovery, we should not have a redelivery. Now I see how the CrashMessage is trying to hold the re-delivery.

File: akka-persistence/src/test/scala/akka/persistence/AtLeastOnceDeliveryCrashSpec.scala [hunk 2]
```diff
@@ -75,6 +77,8 @@ class AtLeastOnceDeliveryCrashSpec
       testProbe.expectMsgType[CrashingActor.SendingMessage]
 
       superVisor ! CrashingActor.CrashMessage
+      testProbe.expectMsgType[CrashingActor.ConfirmCrashMessage.type]
```
Comment (octonato) [inline akka-persistence/src/test/scala/akka/persistence/AtLeastOnceDeliveryCrashSpec.scala:80]: Ok, now that I better understand how this tests is working, I think I found the issue. 

In the logs, I found that `CrashMessage` was sent to dead-letter. That's because `superVisor` stopped before the message arriving and therefore on re-instantiation the actor will never crash and the `SendingMessage` will be redelivery which is not what this tests is expecting. 

Now we wait for a confirmation that the `CrashMessage` was delivered and persisted before we stop `superVisor`

**Review by johanandren — APPROVED** [review]
LGTM!

**Review by raboof — APPROVED** [review]
Looks good!
